checks.d/pgbouncer.py

```diff
@@ -0,0 +1,165 @@
+        AgentCheck.__init__(self, name, init_config, agentConfig, instances)
+        self.dbs = {}
+
+    def _collect_stats(self, key, db, instance_tags):
+        """Query pgbouncer for various metrics
+        """
+
+        metric_scope = [self.STATS_METRICS, self.POOLS_METRICS]
+
+        try:
+            cursor = db.cursor()
+            for scope in metric_scope:
+
+                cols = scope['metrics'].keys()
+
+                try:
+                    query = scope['query']
+                    self.log.debug("Running query: %s" % query)
+                    cursor.execute(query)
+
+                    results = cursor.fetchall()
+                except pg.Error, e:
+                    self.log.warning("Not all metrics may be available: %s" % str(e))
+                    continue
+
+                for row in results:
+                    if row[0] == self.DB_NAME:
+                        continue
+
+                    desc = scope['descriptors']
+                    assert len(row) == len(cols) + len(desc)
+
+                    tags = instance_tags[:]
+                    tags += ["%s:%s" % (d[0][1], d[1]) for d in zip(desc, row[:len(desc)])]
+
+                    values = zip([scope['metrics'][c] for c in cols], row[len(desc):])
+
+                    [v[0][1](self, v[0][0], v[1], tags=tags) for v in values]
+
+            if not results:
+                self.warning('No results were found for query: "%s"' % query)
+
+            cursor.close()
+        except pg.Error, e:
+            self.log.error("Connection error: %s" % str(e))
+            raise ShouldRestartException
+
+    def _get_connection(self, key, host, port, user, password, use_cached=True):
+        "Get and memoize connections to instances"
+        if key in self.dbs and use_cached:
+            return self.dbs[key]
+
+        elif host != "" and user != "":
+            try:
+                service_check_tags = [
+                    "host:%s" % host,
+                    "port:%s" % port
+                ]
+                service_check_tags.append("db:%s" % self.DB_NAME)
+
+                if host == 'localhost' and password == '':
+                    # Use ident method
+                    connection = pg.connect("user=%s dbname=%s" % (user, self.DB_NAME))
+                elif port != '':
+                    connection = pg.connect(host=host, port=port, user=user,
+                        password=password, database=self.DB_NAME)
+                else:
+                    connection = pg.connect(host=host, user=user, password=password,
+                        database=self.DB_NAME)
+                status = AgentCheck.OK
+                self.service_check('pgbouncer.can_connect', status, tags=service_check_tags)
+                self.log.debug('pgbouncer status: %s' % status)
+
+            except Exception:
+                status = AgentCheck.CRITICAL
+                self.service_check('pgbouncer.can_connect', status, tags=service_check_tags)
```

Comment:
You can maybe pass the exception as the service check message to give an idea of what's wrong.